tests/token/erc1155/test_ERC1155MintableBurnable.py

```diff
@@ -0,0 +1,1468 @@
+            *uarr2cd(TRANSFER_AMOUNTS), DATA
+        ])
+
+    execution_info = await erc1155.balanceOfBatch(
+        [sender]*3+[recipient]*3, TOKEN_IDS*2).invoke()
+    assert execution_info.result.balances[:3] == TRANSFER_DIFFERENCES
+    assert execution_info.result.balances[3:] == TRANSFER_AMOUNTS
+
+
+@pytest.mark.asyncio
+async def test_safe_batch_transfer_from_emits_event(erc1155_minted_factory):
+    erc1155, account1, account2, _ = erc1155_minted_factory
+
+    sender = account2.contract_address
+    recipient = account1.contract_address
+
+    execution_info = await signer.send_transaction(
+        account2, erc1155.contract_address, 'safeBatchTransferFrom',
+        [
+            sender, recipient, *uarr2cd(TOKEN_IDS),
+            *uarr2cd(TRANSFER_AMOUNTS), DATA
+        ])
+
+    assert_event_emitted(
+        execution_info,
+        from_address=erc1155.contract_address,
+        name='TransferBatch',
+        data=[
+            sender,     # operator
+            sender,     # from
+            recipient,  # to
+            *uarr2cd(TOKEN_IDS),
+            *uarr2cd(TRANSFER_AMOUNTS),
+        ]
+    )
+
+
+@pytest.mark.asyncio
+async def test_safe_batch_transfer_from_approved(erc1155_minted_factory):
+    erc1155, account1, account2, _ = erc1155_minted_factory
+
+    sender = account2.contract_address
+    operator = account1.contract_address
+    recipient = account1.contract_address
+
+    await signer.send_transaction(
+        account2, erc1155.contract_address, 'setApprovalForAll',
+        [operator, TRUE])
+
+    await signer.send_transaction(
+        account1, erc1155.contract_address, 'safeBatchTransferFrom',
+        [
+            sender, recipient, *uarr2cd(TOKEN_IDS),
+            *uarr2cd(TRANSFER_AMOUNTS), DATA
+        ])
+
+    execution_info = await erc1155.balanceOfBatch(
+        [sender]*3+[recipient]*3, TOKEN_IDS*2).invoke()
+    assert execution_info.result.balances[:3] == TRANSFER_DIFFERENCES
+    assert execution_info.result.balances[3:] == TRANSFER_AMOUNTS
+
+
+@pytest.mark.asyncio
+async def test_safe_batch_transfer_from_approved_emits_event(
+        erc1155_minted_factory):
+    erc1155, account1, account2, _ = erc1155_minted_factory
+
+    sender = account2.contract_address
+    operator = account1.contract_address
+    recipient = account1.contract_address
+
+    await signer.send_transaction(
+        account2, erc1155.contract_address, 'setApprovalForAll',
+        [operator, TRUE])
+
+    execution_info = await signer.send_transaction(
+        account1, erc1155.contract_address, 'safeBatchTransferFrom',
+        [
+            sender, recipient, *uarr2cd(TOKEN_IDS),
+            *uarr2cd(TRANSFER_AMOUNTS), DATA
+        ])
+
+    assert_event_emitted(
+        execution_info,
+        from_address=erc1155.contract_address,
+        name='TransferBatch',
+        data=[
+            operator,   # operator
+            sender,     # from
+            recipient,  # to
```

Comment:
technically, since `operator == recipient` this test would pass even if the event is wrongly emitting the same operator and recipient even when they shouldn't. not sure if it's worth fixing though.